finalfrontier/src/dep_trainer.rs

```diff
@@ -0,0 +1,232 @@
+    dep_config: DepembedsConfig,
+    common_config: CommonConfig,
+    input_vocab: Arc<SubwordVocab>,
+    output_vocab: Arc<SimpleVocab<Dependency>>,
+    rng: R,
+}
+
+impl<R> DepembedsTrainer<ReseedOnCloneRng<R>>
+where
+    R: Rng + Clone + SeedableRng,
+{
+    /// Constructs a new `DepTrainer`.
+    pub fn new(
+        input_vocab: SubwordVocab,
+        output_vocab: SimpleVocab<Dependency>,
+        common_config: CommonConfig,
+        dep_config: DepembedsConfig,
+        rng: R,
+    ) -> Self {
+        DepembedsTrainer {
+            common_config,
+            dep_config,
+            input_vocab: Arc::new(input_vocab),
+            output_vocab: Arc::new(output_vocab),
+            rng: ReseedOnCloneRng(rng),
+        }
+    }
+}
+
+impl<R> DepembedsTrainer<R>
+where
+    R: Rng,
+{
+    fn iter_impl<I>(
+        &mut self,
+        dep_iter: I,
+        sentence: &Sentence,
+    ) -> Box<Iterator<Item = (usize, Vec<usize>)>>
+    where
+        I: Iterator<Item = (usize, Dependency)>,
+    {
+        let invalid_idx = self.input_vocab.len();
+        let mut tokens = vec![invalid_idx; sentence.len() - 1];
+        for (idx, token) in sentence.iter().filter_map(|node| node.token()).enumerate() {
+            if let Some(vocab_idx) = self.input_vocab.idx(token.form()) {
+                if self.rng.gen_range(0f32, 1f32) < self.input_vocab.discard(vocab_idx) {
+                    tokens[idx] = vocab_idx
+                }
+            }
+        }
+
+        let mut contexts = vec![Vec::new(); sentence.len() - 1];
+        for (focus, dep) in dep_iter.filter(|(focus, _dep)| tokens[*focus] != invalid_idx) {
+            if let Some(dep_id) = self.output_vocab.idx(&dep) {
+                if self.rng.gen_range(0f32, 1f32) < self.output_vocab.discard(dep_id) {
+                    contexts[focus].push(dep_id)
+                }
+            }
+        }
+        Box::new(
+            tokens
+                .into_iter()
+                .zip(contexts.into_iter())
+                .filter(move |(focus, _)| *focus != invalid_idx),
+        )
+    }
+}
+
+impl<R> NegativeSamples for DepembedsTrainer<R>
+where
+    R: Rng,
+{
+    fn negative_sample(&mut self, output: usize) -> usize {
+        loop {
+            let negative = self.rng.gen_range(0, self.output_vocab.len());
+            if negative != output {
+                return negative;
+            }
+        }
+    }
+}
+
+impl<R> TrainIterFrom<Sentence> for DepembedsTrainer<R>
+where
+    R: Rng,
+{
+    type Iter = Box<Iterator<Item = (usize, Vec<usize>)>>;
+    type Contexts = Vec<usize>;
+
+    fn train_iter_from(&mut self, sentence: &Sentence) -> Self::Iter {
+        let depth = self.dep_config.depth as usize;
+        let graph = sentence.dep_graph();
+        let dep_iter = DependencyIterator::new(&graph, depth);
+        use crate::dep_trainer::{FilterRootEnum::*, NormalizedEnum::*, TypedEnum::*};
+        match dep_config_to_tuple(&self.dep_config) {
```

Comment:
Not very happy about this part, do you see a nicer way to do it? The training binary needs more or less the same match block to build the vocabulary.

Reply:
Yep, iterator types are great for static polymorphism, bad for dynamic polymorphism. Since there are only four relevant iterator types here, you could consider wrapping them in ad-hoc enum that also implements `DepIter`. (`DepIterBuilder`?)

Then there are two possibilities: implement `Iterator<Item = (usize, Dependency)>` for this ad-hoc enum. Then you get `DepIter` for free, but the downside is that each iterator has a couple of wrapped intermediate wrapper enums as well (though they might be compiled away). Or you only implement the `DepIter` trait, which has the benefit that you can unwrap the enum variant, apply the correct method and then wrap it again in the enum. Finally, this enum could have a method that returns a boxed dependency iterator.

It is more code, but also easier to reduce. This method would then become something like

```rust
let builder = DepIterBuilder::new(dep_iter);
if ... {
    builder.filter_root();
}
if ... {
    builder.normalized();
}

//...

self.iter_impl(builder.into_iter());
```

I was hoping that there was some way to apply a self-consuming map on a `Box`, but unfortunately there does not seem to be such a thing, because of difficulty in computing the requires sizes.

Reply:
BTW. look good otherwise.

Reply:
I didn't end up with a (nice) enum implementation but rather went with a wrapper struct:

```Rust
pub struct DepIterBuilder<'a> {
    inner: Box<Iterator<Item = (usize, Dependency)> + 'a>,
}
```

Construction is simply done in this method, not requiring any conditionals or unwrapping or any other juggling:
```Rust
    pub fn new_from_config(graph: &'a DepGraph<'a>, config: DepembedsConfig) -> Self {
        let mut iter = DepIterBuilder {
            inner: Box::new(DependencyIterator::new(graph, config.depth as usize)),
        };
        if !config.use_root {
            iter.inner = Box::new(iter.inner.filter_root());
        }
        if config.untyped {
            iter.inner = Box::new(iter.inner.untyped());
        }
        if config.normalize {
            iter.inner = Box::new(iter.inner.normalized());
        }
        iter
    }
```

Reply:
But that gives several indirections on the heap for every `next()` call?

Reply:
(And I really disliked the one *necessary* indirection, but this is getting out of hand ;).)

Reply:
I tried to implement the enum way, but I can't make it compile without excessive boxing. 

The first approach failed for me because we end up with the exact same situation as before: After implementing `Iterator<Item = (usize, Dependency)>` on the enum, we get the `DepIter` impl for free, but that is somewhat redundant because we had that anyways on the `DependencyIterator`. Instantiating the enum with some `DepIterBuilder::Base(DependencyIterator)` and calling `.normalize()` on it returns a `Normalized<DepIterBuilder<DependencyIterator>` which cannot be wrapped again in a `DepIterBuilder` because it's a cyclic type with infinite size. Even if there is a workaround for the size issue, we still can't assign a variable of type `DepIterBuilder<Normalized<DepIterBuilder<DependencyIterator>>` to `DepIterBuilder<DependencyIterator>`. 

Which points to how the second approach failed:

Manually implementing `DepIter`on the enum gives us conversion from `DepIterBuilder<I>` to e.g. `Normalized<I>`. If we wrap this `Normalized<I>` in the enum, we end up with `DepIterBuilder<Normalized<I>>`, which again is of a different type.

At this point, I'd suggest we simply add a method (that's just a mock-name)
```Rust
pub fn apply_cfg<'a>(self, config: DepembedsConfig) -> Box<Iterator<Item = (usize, Dependency)> + 'a>
```
and live with having 8 match-arms in this part of the code.

That will not add a lot of indirection (single box) on the heap and it's maintainable since the ugly part is contained in a single method.

What's your take on that suggestion?

Reply:
Ah, yes, the problem is that the inner iterator type is included as a type parameter, to make it work, we have to erase the type (which makes the size undefined).

Ok, go for the function that constructs the proper iterator.